### DISCUSSION TITLE
Don't use Control spacers, instead you can use size flags and BoxContainer's separation for the same result.

### DISCUSSION THREAD

**Discussion by WhalesState** [godotengine/godot-proposals#7444] [Editor]
![image](https://github.com/godotengine/godot-proposals/assets/53877170/1cfc5419-781d-486d-b174-0f8028f578d6)

![image](https://github.com/godotengine/godot-proposals/assets/53877170/ceefa5e3-beaf-40dd-b784-3b1ab288dd1a)

![image](https://github.com/godotengine/godot-proposals/assets/53877170/1df61502-7459-460a-aa38-edb462c0dd08)

![image](https://github.com/godotengine/godot-proposals/assets/53877170/42478f61-a7a8-462f-98a6-e38d14b005d5)

![image](https://github.com/godotengine/godot-proposals/assets/53877170/158b39a7-83d6-4aaa-8133-f7e8b64f0fca)

![image](https://github.com/godotengine/godot-proposals/assets/53877170/e91aac2f-04df-4802-9ad6-1e5ccfae6869)

**Comment by theraot** [comment]
Is this a proposal to change some part of Godot? If so, can you explain what are you modifying? Also, since it seems you hade done the modifications on your end, are you planning to make a pull request?

**Reply by WhalesState** [reply]
I've done the modifications on this branch https://github.com/WhalesState/godot-2D/tree/GLES3, you can have a look at the last 4 commits, also there will be much more spacers in 3d editor plugins so you can track `add_spacer(`, `spacing`, `spacer` and `space`. then we can set `add_spacer` from `BoxContainer` as deprecated since it's not efficient to use it.

**Reply by dalexeev** [reply]
> then we can set `add_spacer` from `BoxContainer` as deprecated since it's not efficient to use it.

Duplicate my comment from issue:

Note that this is only true for two children. If the container has 3 or more children of different sizes (and use 2 or more spacers between them), then the same behavior as with spacers cannot be achieved using size flags, as far as I know.

**Comment by theraot** [comment]
Apparently these are multiple files being modified:

- godot/editor/animation_track_editor.cpp

    ![](https://user-images.githubusercontent.com/53877170/258388242-1cfc5419-781d-486d-b174-0f8028f578d6.png)

- godot/editor/connections_dialog.cpp

    ![](https://user-images.githubusercontent.com/53877170/258388366-ceefa5e3-beaf-40dd-b784-3b1ab288dd1a.png)

- godot/editor/editor_plugin_settings.cpp

    ![](https://user-images.githubusercontent.com/53877170/258388557-1df61502-7459-460a-aa38-edb462c0dd08.png)

- godot/editor/groups_editor.cpp

    ![](https://user-images.githubusercontent.com/53877170/258388632-42478f61-a7a8-462f-98a6-e38d14b005d5.png)

- godot/editor/import_dock.cpp

    ![](https://user-images.githubusercontent.com/53877170/258388767-158b39a7-83d6-4aaa-8133-f7e8b64f0fca.png)
    
- godot/editor/project_manager.cpp

    ![](https://user-images.githubusercontent.com/53877170/258388862-e91aac2f-04df-4802-9ad6-1e5ccfae6869.png)

*Yes, I transcribed parts of the pictures and searched them in the code base. Yes, this is not an efficient way of doing things.*

**Comment by WhalesState** [comment]
if you need any help please consider converting this to an issue then create a branch and give me access to it so I can apply the modifications from my branch and help removing the others because I will need some help doing it in some scripts since i can't track them if they are 3d related or with editor debugger if they are not inside the editor window.

**Comment by WhalesState** [comment]
based on this latest [commit](https://github.com/WhalesState/godot-2D/commit/b4af47a2abef1261f1073dda051a2e944d8177d1) to the GLES3 branch I have removed more than 100 control nodes from the editor, and also have set `BoxContainer.add_spacer()` as deprecated, the remarkable edit that has eliminated the most of them was done to the Dialog buttons.

This enhancement is particularly beneficial for the Editor, as it makes extensive use of Dialogs. By removing these control nodes, I have significantly improved the performance of the Editor.

Control nodes, especially when large in size, can be resource-intensive for both the GPU and CPU. and you can improve the add_spacer function with just one line of code that prevents it's fill vertically if it's a horizontal spacer or vise-verse.

**Reply by Zireael07** [reply]
> Control nodes, especially when large in size, can be resource-intensive for both the GPU and CPU

Especially so on iOS devices going by #78278

**Reply by WhalesState** [reply]
I feel bad working on this 2d project since many people are attacking me thinking i will make the devs spirit low and someone have said that this may result in dropping support for 2d in godot, while all i want is to backport all those changes to godot for a better 2d support, the 3d code was so much confusing for a noob like me and to tell which code is used for 3d and which is used for 2d was so hard since there are many cases that the same code is used for both or a 2d class depends mainly on it's 3d equivalent class. also i admire the devs so much and I'm not planning to include my name in my branches, i started it to have a base for my 2d projects and i always give full credits to godot engine. thank you for making such wonderful engine and hope i get better in c++ soon so i can help efficiently.

**Comment by WhalesState** [comment]
here's a benchmark to prove how the changes i have made can benefit godot when they gets implemented, just if anyone is ready to help me we can start on optimizing godot 2d builds and the editor, thank you ^^

![GLES](https://github.com/godotengine/godot-proposals/assets/53877170/ab581794-b856-424c-88df-c470f033e5df)

**Reply by WhalesState** [reply]
note that Godo version was built with `optimize=speed_trace`, the name change is to prevent confusing or editor data folder overwrite.

**Reply by timothyqiu** [reply]
Is this benchmark comparing before & after the removal of spacers, or comparing vanilla Godot & your godot-2D project?

**Reply by KoBeWi** [reply]
It seems to be the latter. Removing spacers don't make much difference (if at all).

**Reply by WhalesState** [reply]
this comparison is godot 4.1.1 stable on bottom as the default godot running editor and game using Gles3 and running the same scene and opening the scene in CanvasItemEditor, and on the other hand [godot-2d](https://github.com/WhalesState/godot-2D/actions/runs/5775807524) from this action build on github, this is not only showing the benefits of removing Control spacers but also allowing the editor to be built with 3d_disabled and disabling more 3d code in 2d builds which will give better benchmark for godot 4.x when they do comparison for other engines or for 3.x and 4.x, i will try to see the differences before and after removing the control spacers to see if it makes changes to the editor CPU and GPU usage.

**Reply by WhalesState** [reply]
> It seems to be the latter. Removing spacers don't make much difference (if at all).

Do you mean that running a scene with +100 control nodes with rect_size(windows_size.x, button_size.y) is the same as running an empty scene with no much difference? even when most of them are not visible on screen they still exist and referenced inside the editor affecting it's nodes count and memory usage!

Every small optimization is recommended especially when we do things correctly and efficiently. at least for the users who are using godot always alongside other apps or browsers. have a look on this issue https://github.com/godotengine/godot/issues/78278

**Comment by WhalesState** [comment]
this is what happened after removing the Control spacers for both editor builds with 3d_disabled and more disabling of 3d code, the comparison is between this version which is the first one in the image [after removing spacers](https://github.com/WhalesState/godot-2D/actions/runs/5775807524) and this one [before removing spacers](https://github.com/WhalesState/godot-2D/actions/runs/5749396843), actually removing them has resulted in a change for both CPU and GPU for the editor, and you can do better comparison than mine.
![GLES](https://github.com/godotengine/godot-proposals/assets/53877170/c851ba4f-5708-4753-9f69-456a26e181e0)

**Reply by AThousandShips** [reply]
I think it should be obvious that something else is causing the 80% drop in RAM usage between the two tests, the conditions for the test are not equivalent, something open in the editor differing

**Comment by WhalesState** [comment]
> I think it should be obvious that something else is causing the 80% drop in RAM usage between the two tests, the conditions for the test are not equivalent, something open in the editor differing

both versions was built on github using the same flags, also I will not achieve anything by sharing fake info ^^ all i want is helping but anyway, thanks for your kind words on my hard work. if something is wrong here then it's not my fault, same scene and same game is running and you can tell from the names.
![image](https://github.com/godotengine/godot-proposals/assets/53877170/1c94f199-57fc-4c5b-ac2c-c4d435ee5637)

**Reply by AThousandShips** [reply]
> also I will not achieve anything by sharing fake info 

Nor have I said that you have, if I thought that I would have taken a very different approach to you, I do believe this is an honest mistake
